isabelle/Core/SemanticsTypes.thy

```diff
@@ -367,29 +258,79 @@ and Contract = Close
+a payee \secref{sec:internal-accounts} \<^term>\<open>p\<close> of \#\<^term>\<open>v\<close> \<^term>\<open>Token\<close>s and then continues to contract \<^term>\<open>c\<close>.
+Warnings will be generated if the value @{term v} is negative, or if there is not enough in the account to
+make the payment in full. In the latter case, a partial payment (of the available amount) is made
+\<close>
+
+text \<open>The contract \<^term>\<open>If obs x y\<close> allows branching. We continue to branch \<^term>\<open>x\<close> if the \<^term>\<open>Observation obs\<close>
+evaluates to \<^term>\<open>true\<close>, or branch\<^term>\<open>y\<close> otherwise.
+\<close>
+
+text \<open>@{term When} is the most complex constructor for contracts, with the form
+@{term "When cs t c"}. The list \<^term>\<open>cs\<close> contains zero or more pairs of \<^term>\<open>Action\<close>s and
+ \<^term>\<open>Contract\<close> continuations. When we do a \<^term>\<open>computeTransaction\<close> \secref{sec:computeTransaction},
+ we follow the continuation associated to the first \<^term>\<open>Action\<close> that matches the \<^term>\<open>Input\<close>. If no
+ action is matched it returns a \<^term>\<open>ApplyAllNoMatchError\<close>. If a valid \<^term>\<open>Transaction\<close> is computed
+ with a \<^term>\<open>TimeInterval\<close> with a start time bigger than the \<^term>\<open>Timeout t\<close>, the contingency
+ continuation \<^term>\<open>c\<close> is evaluated. The explicit timeout mechanism is what allow Marlowe to avoid
+waiting forever for external inputs.\<close>
+
+text \<open>A @{term Let} contract @{term "Let i v c"} allows a contract to name a
+value using an identifier @{term i}. In this case, the expression @{term v} is
+evaluated, and stored with the name @{term i}. The contract then continues as
+@{term c}. As well as allowing us to use abbreviations, this mechanism also means that we
+can capture and save volatile values that might be changing with time, e.g. the current price of
+oil, or the current time, at a particular point in the execution of the contract, to be used later
+on in contract execution.
+\<close>
+
+text \<open>An assertion contract @{term "Assert b c"} does not have any effect on
+the state of the contract, it immediately continues as @{term c}, but it issues a warning
+when the observation @{term b} is false. It can be used to ensure that a property holds
+in any given point of the contract, since static analysis will fail if any execution causes a
+warning. the @{term Assert} term might be removed from future on-chain versions of Marlowe.\<close>
+
+
+subsection \<open>State and Environment \label{sec:state-and-env}\<close>
+
+text \<open>The internal state of a Marlowe contract consists of the current balances in each party's
+account, a record of the most recent value of each type of choice, a record of the most recent
+value of each variable, and the minimum time for which input can be applied to the contract. The
```

Comment:
```suggestion
value of each variable, and the lower bound for the current time that is used to refine time intervals and ensure @{term TimeIntervalStart} never decreases. The
```
It is possible to apply transactions with intervals that have the lower bound smaller than minTime, but they get truncated to minTime